snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+	}
+
+	@FunctionalInterface
+	public interface ConceptSetEvaluator {
+		
+		ConceptSetEvaluator DEFAULT = (context, locales, resourceUri, inclusions, exclusions, pageSize) -> {
+			// No inclusions, return early with empty result
+			if (inclusions.isEmpty()) {
+				return newHashSet();
+			}
+			
+			Set<String> queryExpressions = inclusions.stream()
+				.map(QueryExpression::getQuery)
+				.collect(Collectors.toSet());
+			
+			if (queryExpressions.isEmpty()) {
+				queryExpressions = null;
+			}
+			
+			Set<String> mustNotQueryExpressions = exclusions.stream()
+				.map(QueryExpression::getQuery)
+				.collect(Collectors.toSet());
+
+			if (mustNotQueryExpressions.isEmpty()) {
+				mustNotQueryExpressions = null;
+			}
+
+			return CodeSystemRequests.prepareSearchConcepts()
+				.filterByCodeSystemUri(resourceUri)
+				.filterByInclusions(queryExpressions)
+				.filterByExclusions(mustNotQueryExpressions)
+				.setLimit(PAGE_SIZE)
+				.setLocales(locales)
+				.stream(context)
+				.flatMap(Concepts::stream)
+				.map(Concept::getId)
+				.collect(Collectors.toSet());
+		};
+		
+		Set<String> evaluateConceptSet(
+			BranchContext context, 
+			List<ExtendedLocale> locales, 
+			ResourceURI resourceUri, 
+			Collection<QueryExpression> inclusions, 
+			Collection<QueryExpression> exclusions, 
+			int pageSize);
+	}
+
+	// External dependencies, can be modified for testing purposes
+	private Clock clock = Clock.systemUTC();
+	private EclEvaluator evaluator = EclEvaluator.DEFAULT;
+	private EclLabeler labeler = EclLabeler.DEFAULT;
+	private ConceptSetEvaluator conceptSetEvaluator = ConceptSetEvaluator.DEFAULT;
+
+	private RelationshipSearchBySource relationshipSearchBySource = RelationshipSearchBySource.DEFAULT;
+	private RelationshipSearchByTypeAndDestination relationshipSearchByTypeAndDestination = RelationshipSearchByTypeAndDestination.DEFAULT;
+
+	private ConceptSearchById conceptSearchById = ConceptSearchById.DEFAULT;
+	private ConceptDescendantCountById conceptDescendantCountById = ConceptDescendantCountById.DEFAULT;
+	private EdgeSearchBySourceId edgeSearchBySourceId = EdgeSearchBySourceId.DEFAULT;
+
+	private ResourceURI resourceUri;
+	private Logger log;
+	
+	// Optimizer parameters (some of these are adjusted dynamically)
+	private final int minimumClusterSize       = 2;
+	private final float falsePositiveThreshold = 0.7f;
+	private final double weight                = 1.0d;
+	private final float minimumFitThreshold    = 0.5f;
+	private float fitThreshold                 = 0.9f;
+	private final int idealClauseCount         = 1;
+	private final int maxIteration             = 1000;
+	private final Duration maxRuntime          = Duration.ofSeconds(580L);
+
+	private OptimizerStrategy optimizerStrategy = OptimizerStrategy.DEFAULT;
+	private int zoom                            = 1;
+	private int maxClauseCount                  = 2000;
+
+	// Run-time fields
+	private Set<String> conceptSet;
```

Comment:
This is a server-side singleton instance in the current implementation (see `SnomedPlugin`). Running multiple optimizations at the same time will mess up the internals. Please refactor. Feel free to create a new instance per request.